porch/internal/api/porchinternal/v1alpha1/types.go

```diff
@@ -33,6 +33,7 @@ type PackageRev struct {
 
 // PackageRevSpec defines the desired state of PackageRev
 type PackageRevSpec struct {
+	DeletionProposed bool `json:"deletionProposed,omitempty"`
```

Comment:
Thanks for pointing this out. I had similar reservations about storing it in the CRD, but agreed that it was not obvious how to store it in git and oci. I think there are a few options for that:

1. Storing the package revision lifecycle in the commit annotations for git, and in tags for oci. Implementation-wise, that would mean iterating over the commits to find the latest one corresponding to a particular revision to read the lifecycle value. I think we already have code that does this, or something similar. I'm not that familiar with how oci works though, but it's probably doable.

2. Creating a new task type that represents the lifecycle going from Published->DeletionProposed or vice versa. Replaying the tasks should land the package revision in the correct lifecycle in this case. But I recall that we want to be cautious about adding too many different types of tasks.

3. Putting the lifecycle in the Kptfile, maybe under the `status` field. This might be an odd thing to have when users use kpt without porch, e.g. if they do a `kpt pkg get` on a package that has a lifecycle field in the Kptfile, we'd need to decide whether it makes sense to clear the lifecycle field or keep it.

I think the first option might be best, wydt?

Reply:
Yeah, I can think of two options here. The first is the same as you mentioned, which is to include it in the commit annotations. The other option is to see if there is a way to leverage tags/branches similar to how we handle the other lifecycle statuses, but in a way that doesn't break published packages. Maybe we create a proposedDelete branch from the tag of the published package, but without changing the published tag?

The concern about using messages in git commits is sometimes that it doesn't work well with packages created outside of porch, but I don't think that applies in this case. The lifecycle of packages are inherently tied to porch and some scheme using a branch/tag also would also only be useful through porch.

Reply:
If we were doing gitops directly, I would expect to see a PR that deletes all the files in the package (maybe leaving a tombstone if we decide we need one to avoid problems of package / package-version reuse).

I would suggest if we can treat it like other changes, that would be good therefore.  So in git/oci we have a change branch, there's some metadata in the commit message or layer metadata that helps us figure out what it is without having to reverse engineer the change, and if the draft is approved then we create a new version of the package (in this case, a quite weird version of the package where it is empty or deleted; if we need a tombstone that's fine but I suspect the metadata will be sufficient)

So I would ideally like branches that are a draft like any other, and I think we extend the metadata - I don't _think_ deletion is a task but it could be represented as an empty task-list (if that's not too weird).

I think if merged there would be a new version of the package (a tombstone _version_), because otherwise it's hard to know that we are expected to update to delete a package (and I don't think we want to just accidentally leave those packages in whatever state they happen to be in).  Thinking about "deletion proposed" as a draft PR of that final revision might be another good way to think about it!

Reply:
> Possibly updating the revision of the package on the main branch. The meaning of the main branch has been a bit strange since it is always the last published revision that is reflected there, but this is clearer now that the last published revision must also be the latest revision. So if we delete the latest published revision, we should ideally also update the revision of the package on the main branch to reflect what is the latest published revision after the deletion.

It seems like today, we don't actually do this yet where we roll back to the previous revision after deleting the latest published revision, but I agree that ideally we would.

If we wanted to create a branch to represent the proposed deletion, I think we'd want it to look like what the main branch would look like after deletion. That means if it's the latest published revision being deleted, it should have the package look like the previous revision; if we're deleting some other published revision, then the branch would look identical to head. If more packages get published in the meantime, this branch would fall behind and not be correct anymore, unless we periodically check and update it. This option probably looks the most like merging a PR, but at the end of the day we probably wouldn't merge the branch; we'd just be using the existence of the branch as a boolean, in which case keeping the deletion branch up to date doesn't seem worth it to me. Since it's just a boolean we need to store (whether or not the revision has been proposed for deletion), I think I prefer to store that boolean in the commit message rather than creating a new branch that might eventually become stale.